### PR TITLE
State: Improve schema violation warning

### PR DESCRIPTION
**Pull request by sirreal**
This PR improves messaging produced by state schema violations. These changes are only applied when _not in_ production environment.

1. Use `greedy` mode to continue validation after an error is encountered. This may produce several violations at once.
1. Use `verbose` mode to get `schemaPath`, allowing us to point to the specific rule violated.
1. Some formatting updates.

Example message (from console test output):

```
      State validation failed.
      State: { jank: 'y', isAuthorizing: 'maybe' }

      data.timestamp is required
      Found: { jank: 'y', isAuthorizing: 'maybe' }

      data has additional properties
      Found: 'data.jank'

      data.isAuthorizing must be an enum value
      Found: 'maybe'
      Violates rule: { type: 'string', enum: [ 'a', 'b', [length]: 2 ] }
```

## Testing
* Create a schema violation and observe it in the console.
  - In the browser, create a schema violation in persisted state and refresh
  - Apply the patch below and run
    ```sh
    npm run test-client client/state/test/persistence.js
    ```

### testing patch

```diff
diff --git a/client/state/jetpack-connect/reducer/jetpack-connect-authorize.js b/client/state/jetpack-connect/reducer/jetpack-connect-authorize.js
index fdfa63e52..801585ea6 100644
--- a/client/state/jetpack-connect/reducer/jetpack-connect-authorize.js
+++ b/client/state/jetpack-connect/reducer/jetpack-connect-authorize.js
@@ -28,7 +28,10 @@ import {
 	SITE_REQUEST_FAILURE,
 } from 'state/action-types';
 
-export default function jetpackConnectAuthorize( state = {}, action ) {
+export default function jetpackConnectAuthorize(
+	state = { jank: 'y', isAuthorizing: 'maybe' },
+	action
+) {
 	switch ( action.type ) {
 		case JETPACK_CONNECT_AUTHORIZE:
 			return Object.assign( omit( state, 'userData', 'bearerToken' ), {
diff --git a/client/state/jetpack-connect/reducer/schema.js b/client/state/jetpack-connect/reducer/schema.js
index f12ade244..bd1a21932 100644
--- a/client/state/jetpack-connect/reducer/schema.js
+++ b/client/state/jetpack-connect/reducer/schema.js
@@ -1,32 +1,23 @@
 /** @format */
 export const jetpackConnectAuthorizeSchema = {
-	anyOf: [
-		{
-			type: 'object',
-			additionalProperties: false,
-			properties: {},
-		},
-		{
-			type: 'object',
-			additionalProperties: true,
-			required: [ 'timestamp' ],
-			properties: {
-				authorizationCode: { type: 'string' },
-				authorizeError: { type: [ 'boolean', 'null' ] },
-				authorizeSuccess: { type: 'boolean' },
-				bearerToken: { type: 'string' },
-				clientId: { type: 'integer' },
-				clientNotResponding: { type: 'boolean' },
-				isAuthorizing: { type: 'boolean' },
-				isRedirectingToWpAdmin: { type: 'boolean' },
-				plansUrl: { type: 'string' },
-				siteReceived: { type: 'boolean' },
-				timestamp: { type: 'integer' },
-				userAlreadyConnected: { type: 'boolean' },
-				userData: { type: 'object' },
-			},
-		},
-	],
+	type: 'object',
+	additionalProperties: false,
+	required: [ 'timestamp' ],
+	properties: {
+		authorizationCode: { type: 'string' },
+		authorizeError: { type: [ 'boolean', 'null' ] },
+		authorizeSuccess: { type: 'string', enum: [ 'a', 'b' ] },
+		bearerToken: { type: 'string' },
+		clientId: { type: 'integer' },
+		clientNotResponding: { type: 'boolean' },
+		isAuthorizing: { type: 'string', enum: [ 'a', 'b' ] },
+		isRedirectingToWpAdmin: { type: 'boolean' },
+		plansUrl: { type: 'string' },
+		siteReceived: { type: 'boolean' },
+		timestamp: { type: 'integer' },
+		userAlreadyConnected: { type: 'boolean' },
+		userData: { type: 'object' },
+	},
 };
 
 export const jetpackAuthAttemptsSchema = {
diff --git a/client/state/test/persistence.js b/client/state/test/persistence.js
index 4c81433d7..7a5bbb664 100644
--- a/client/state/test/persistence.js
+++ b/client/state/test/persistence.js
@@ -10,15 +10,11 @@ describe( 'persistence', () => {
 		const consoleErrorSpy = jest
 			.spyOn( global.console, 'error' )
 			.mockImplementation( () => () => {} );
-		const consoleWarnSpy = jest
-			.spyOn( global.console, 'warn' )
-			.mockImplementation( () => () => {} );
 
 		const initialState = createReduxStore().getState();
 
 		reducer( reducer( initialState, { type: SERIALIZE } ), { type: DESERIALIZE } );
 
 		expect( consoleErrorSpy ).not.toHaveBeenCalled();
-		expect( consoleWarnSpy ).not.toHaveBeenCalled();
 	} );
 } );
```

### CODE REVIEW

File: client/state/utils.js [hunk 1]
```diff
@@ -37,14 +40,18 @@ export function isValidStateWithSchema( state, schema, debugInfo ) {
...
 		warn(
-			'state validation failed\nfor state:',
+			'state validation failed for state:\n',
 			state,
-			'\nagainst schema:',
-			schema,
-			'\nwith reason:',
-			prettifyArgument( validate.errors ),
-			'\nsource:',
-			prettifyArgument( debugInfo ) || '(none given)'
```
Comment (nylen) [inline client/state/utils.js]: Let's keep `debugInfo`: it's currently the only way to get Chrome devtools to tell you which reducer this error originates from, even though it's not pretty.

File: client/state/utils.js [hunk 2]
```diff
@@ -37,14 +40,18 @@ export function isValidStateWithSchema( state, schema, debugInfo ) {
...
+			'\nvalidation failures:\n',
+			validate.errors
+				.map(
+					( { field, value, message, type, schemaPath } ) => `
+${ field } ${ message }.
+Found: ${ JSON.stringify( value ) }.
+Schema violation: ${ JSON.stringify( get( schema, schemaPath ) ) }
```
Comment (nylen) [inline client/state/utils.js]: This is sometimes `undefined`.

**Comment by nylen** [comment]
Elaborating on review comments above, here is the purpose that `debugInfo` serves (from #21019):

> 

And the CI output from this PR provides an example of `Schema violation: undefined`:

>

File: client/state/utils.js [hunk 3]
```diff
@@ -27,7 +27,10 @@ import { DESERIALIZE, SERIALIZE } from './action-types';
...
 
 export function isValidStateWithSchema( state, schema, debugInfo ) {
-	const validate = validator( schema );
+	const validate = validator( schema, {
+		greedy: process.env.NODE_ENV !== 'production',
+		verbose: process.env.NODE_ENV !== 'production',
+	} );
 	const valid = validate( state );
 	if ( ! valid && process.env.NODE_ENV !== 'production' ) {
 		function prettifyArgument( obj ) {
```
Comment (nylen) [inline client/state/utils.js]: If this function isn't being used any more (probably not needed with the improved error reporting), let's remove it.

**Comment by sirreal** [comment]
Thanks for the input @nylen! This really wasn't ready to be looked at 🙈 , but I'll keep the input in mind as I try improve these messages. Feel free to push changes if you're so inclined 👍

**Comment by sirreal** [comment]
Rebased

**Comment by gwwar** [comment]
👍  Changes here seem reasonable, feel free to ship after the conflict is resolved.

One thing to note is that we likely won't catch too many issues from the updates to the initialState test. With a well designed reducer, most initial state will be very simple. For example, null, an empty object or an empty array.

**Comment by sirreal** [comment]
Improved debug statements may play nicely with #21125

**Pull request merged**